gallery/__init__.py

```diff
@@ -791,7 +797,14 @@ def display_dir_thumbnail(dir_id: int, auth_dict: Optional[Dict[str, Any]] = Non
         thumbnail_uuid = thumbnail_uuid.split('.')[0]
 
     link = storage_interface.get_link("thumbnails/{}".format(thumbnail_uuid))
-    return redirect(link)
+    if "LOCAL_STORAGE_PATH" in app.config:
+        link = "http://" + app.config["SERVER_NAME"] + link
+    req = requests.get(link)
+    if req.status_code == requests.codes.ok:
+        return req.content
+    dir_model.thumbnail_uuid = refresh_directory_thumbnail(dir_model)
+    db.session.commit()
+    return display_dir_thumbnail(dir_id)
```

Comment:
instead of returning display_dir_thumbnail and doing auth and the db lookup again, why not put this in a loop?